### PR TITLE
Changes for your PR.

### PR DESCRIPTION
**Pull request by myronmarston**
Hey @bf4 you asked for help with the matcher so here you go.  A couple things to note:
- I'm a little bit confused by the encoding check the matcher does.  I ported your logic as-is, which, except for the one spec that passed an explicit encoding,  checks that `actual.encoding == actual.encoding`. (Compare to how it was: an arg list of `str1, str2, expected_encoding=str1.encoding` and `str1.encoding == expected_encoding`).  I tried to change it to check that `actual.encoding == expected.encoding` but that makes some specs fail -- I think that reflects a bug in the specs and/or implementation?  I just left a TODO for it for now.
- Providing a way for the caller to specify the encoding is simple with `chain`, but it seems odd to me to support that...shouldn't the caller just provide the expected string in the expected encoding rather than providing the expected string in one encoding and the encoding as a separate argument?  When the caller forces the encoding, `be_identical_string` no longer seems like an accurate description of the matcher; it's more like `be_identical_string_except_for_the_encoding(string).and_actually_use_this_encoding_instead(encoding)`.  I think we could remove the support for the caller to specify the encoding and just make the caller pass the string in the desired encoding instead, right?  `force_encoding` could be used by the caller to make sure it's right if needed.
- Your other PR had more logic for what has become this matcher that converted to codepoints/bytes for the failure message, right?  If you want to add back that logic, it could go in `failure_message` but I think what we have now is probably sufficient, honestly (besides my concerns above).
- I also corrected a couple comments that I thought were inaccurate/confusing -- let me know what you think.

Since I've targeted the branch that is the basis of your PR, I think that your PR will be updated automatically when you merge this.  Alternately, if you want to cherry-pick commits across and/or do this work in your PR branch yourself, feel free.

Thanks!

### CODE REVIEW

File: spec/rspec/support/encoded_string_spec.rb [hunk 1]
```diff
@@ -2,6 +2,44 @@
...
+    end
+  else
+    match do |actual|
+      actual.split(//) == expected.split(//)
+    end
+  end
+
+  chain :with_encoding do |enc|
+    @expected_encoding = enc
+  end
```
Comment (bf4) [inline spec/rspec/support/encoded_string_spec.rb:32]: Awesome, that's the part I wasn't sure how to do. I couldn't figure out the scoping or syntax.

File: lib/rspec/support/encoded_string.rb [hunk 2]
```diff
@@ -2,9 +2,7 @@ module RSpec
   module Support
     # @private
     class EncodedString
-      # EncodedString is a hotspot in RSpec as every expectation creates
-      # one. Reduce allocations by storing constants.
-      #
+      # Reduce allocations by storing constants.
       UTF_8 = "UTF-8"
       US_ASCII = 'US-ASCII'
```
Comment (bf4) [inline lib/rspec/support/encoded_string.rb:7]: Wow, I am really inconsistent with quoting.

**Comment by bf4** [comment]
Thanks for sticking with me on this.  I know it's been long and frustrating but I think well-worth it in the end.

File: spec/rspec/support/encoded_string_spec.rb [hunk 3]
```diff
@@ -209,9 +246,9 @@ module RSpec::Support
 
           it 'makes no changes to the resulting string' do
             resulting_array = build_encoded_string(non_ascii_compatible_string).split("\n")
-            expect(resulting_array.size).to eq(1) # sanity check
-            expected_string = non_ascii_compatible_string
-            expect_identical_string(resulting_array.first, expected_string, Encoding.find('UTF-16LE'))
+            expect(resulting_array).to match [
```
Comment (bf4) [inline spec/rspec/support/encoded_string_spec.rb:249]: Why do you call `match` here? Would you do that in other projects that use RSpec?

Comment (myronmarston) [inline spec/rspec/support/encoded_string_spec.rb:249]: `match` is a matcher that supports matching array/hash data structures using embedded matchers to match against items in the actual array or hash. See:

http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#_matcher_can_be_used_for_data_structures

In this case, it allows us to assert that `result_array` is an array of one element, which matches the `a_string_identical_to` matcher (defined as an alias of `be_identical_string` above), all in a single expression -- that way, you don't have to do the separate `expect(resulting_array.size).to eq(1)` sanity check.

**Comment by bf4** [comment]
Closed by https://github.com/bf4/rspec-support/commit/34a98d3d44ce433a5e94584ee6fae16b0480808d

**Comment by bf4** [comment]
Closed by 34a98d3

**Comment by bf4** [comment]
I manually merged this, amending your last commit, but it didn't register here for some reason.

**Pull request closed without merging**